# HW9
# Peer-Graded Section

In [1]:
import numpy as np
import pandas as pd
import math as math
import statistics
import matplotlib.pyplot as plt

from numpy.linalg import svd
from sklearn import datasets
from sklearn.decomposition import PCA, NMF

## Question a

In [2]:
movie_review_tibble = pd.read_csv("MovieReviewMat.csv",index_col=0, low_memory=False)
movie_review_tibble

Toy Story (1995)              Jumanji (1995)  \
1     Adventure|Animation|Children|Comedy|Fantasy  Adventure|Children|Fantasy   
2                                             3.5                         3.5   
3                                             2.5                         NaN   
4                                               5                         4.5   
5                                             4.5                           4   
...                                           ...                         ...   
2667                                            3                         2.5   
2668                                            4                           1   
2669                                            3                           3   
2670                                          4.5                         NaN   
2671                                          4.5                         2.5   

     Grumpier Old Men (1995) Waiting to Exhale (1995)  \
1             Comedy|Romance     Comedy|Drama|Romance   
2                          3                      NaN   
3                        NaN                      NaN   
4                        NaN                      NaN   
5                        NaN                      NaN   
...                      ...                      ...   
2667                     1.5                      NaN   
2668                     NaN                      NaN   
2669                     3.5                      NaN   
2670                     NaN                      NaN   
2671                     0.5                        2   

     Father of the Bride Part II (1995)            Heat (1995)  \
1                                Comedy  Action|Crime|Thriller   
2                                   NaN                    NaN   
3                                   NaN                      3   
4                                   NaN                      5   
5                                   NaN                      4   
...                                 ...                    ...   
2667                                NaN                    4.5   
2668                                NaN                    3.5   
2669                                  2                    NaN   
2670                                NaN                      3   
2671                                NaN                    4.5   

      Sabrina (1995) Tom and Huck (1995) Sudden Death (1995)  \
1     Comedy|Romance  Adventure|Children              Action   
2                NaN                 NaN                 NaN   
3                NaN                 NaN                 NaN   
4                NaN                 NaN                 NaN   
5                NaN                 NaN                   3   
...              ...                 ...                 ...   
2667               2                 NaN                   3   
2668               3                 NaN                 NaN   
2669             NaN                 NaN                 NaN   
2670             NaN                 NaN                 NaN   
2671               2                 NaN                 NaN   

               GoldenEye (1995)  ... Madame Sin (1972) Illegally Yours (1988)  \
1     Action|Adventure|Thriller  ...          Thriller         Comedy|Romance   
2                           NaN  ...               NaN                    NaN   
3                             3  ...               NaN                    NaN   
4                             4  ...                 4                    NaN   
5                           3.5  ...               NaN                    NaN   
...                         ...  ...               ...                    ...   
2667                        3.5  ...               4.5                    NaN   
2668                        NaN  ...               NaN                    NaN   
2669                          4  ...               NaN                    NaN   
267

In [3]:
movie_review_names = np.array(movie_review_tibble.columns)
movie_review_genre = np.array(movie_review_tibble.iloc[0])
movie_review = np.array(movie_review_tibble.drop([1]), dtype=float)
movie_review

array([[3.5, 3.5, 3. , ..., nan, nan, nan],
       [2.5, nan, nan, ..., nan, nan, nan],
       [5. , 4.5, nan, ..., nan, nan, nan],
       ...,
       [3. , 3. , 3.5, ..., nan, nan, nan],
       [4.5, nan, nan, ..., nan, nan, 1. ],
       [4.5, 2.5, 0.5, ..., nan, nan, nan]])

Based on the results below, the singular highest and lowest review given to any film is 5 and 0.5, respectively.<br>

The average review is about 3.30. Note that this average is also computed as `np.mean(np.nanmean(movie_review,axis=1))` - where `axis=1` means we're taking the average for each user review (across all movies that they did leave a review) first, then taking the mean across 2671 of such users. If we had done `axis=0`, we would be taking the average review **across all (2153) films** - which results in a slightly different number.

In [4]:
print("Highest review given = " + str(np.nanmax(movie_review)))
print("Lowest review given = " + str(np.nanmin(movie_review)))
print("Overall average review (across 2671 users) = " + str(np.nanmean(movie_review)))

Highest review given = 5.0
Lowest review given = 0.5
Overall average review (across 2671 users) = 3.297080189685473


## Question b
### User 1462
For this user, there are 11 movies which they rate the highest possible score (5) - see below for the names of all 11 movies. <br>Across these 11 movies, the most common genre is **Drama** - it is present in all 11 movies

In [5]:
user1462 = movie_review[1460] #user 1462
index_review_1462 = np.where(user1462 == 5)[0] #len = 11

for i in index_review_1462:
    movie = movie_review_names[i]
    print(movie + " Genre:" + movie_review_tibble.loc[1,movie]) #print name of movies along with genre

Taxi Driver (1976) Genre:Crime|Drama|Thriller
Rob Roy (1995) Genre:Action|Drama|Romance|War
Little Princess, A (1995) Genre:Children|Drama
William Shakespeare's Romeo + Juliet (1996) Genre:Drama|Romance
Quiet Man, The (1952) Genre:Drama|Romance
Raging Bull (1980) Genre:Drama
Pink Floyd: The Wall (1982) Genre:Drama|Musical
Field of Dreams (1989) Genre:Children|Drama|Fantasy
Man Who Would Be King, The (1975) Genre:Adventure|Drama
Playing by Heart (1998) Genre:Drama|Romance
War of the Worlds, The (1953) Genre:Action|Drama|Sci-Fi


### User 45
For this user, there are 5 movies which they rate the highest possible score (5) - see below for the names of all 5 movies. <br>Across these 5 movies, the most common genre is **Comedy** - it is present in all 11 movies. 

**Differences**: User 45 second-most common genre is Drama - which shares with User 1462. On the contrary, user 1462 doesn't have any Comedy movie in their list.

In [6]:
user45 = movie_review[43] #user 45
index_review_5 = np.where(user45 == 5)[0] #len = 5

for i in index_review_5:
    movie = movie_review_names[i]
    print(movie + " Genre:" + movie_review_tibble.loc[1,movie]) #print name of movies along with genre

Jack (1996) Genre:Comedy|Drama
Sting, The (1973) Genre:Comedy|Crime
Groundhog Day (1993) Genre:Comedy|Fantasy|Romance
Howard the Duck (1986) Genre:Adventure|Comedy|Sci-Fi
Anywhere But Here (1999) Genre:Comedy|Drama


## Question c
In terms of genre preference, I expect to see around as many types as there are genres. Like in question b and c, we can choose to look at "preferences" as the most common genre across an individual's most highly-rated movies. In this case, one individual can only have at most one "genre preference".

If we take a look at the first row of the tibble (containing movie genres) and the answers in question b, there seems to be 19 in total (Action, Adventure, Animation, Children, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, IMAX, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western). So I'd expect to see around this many types of of individuals.


## Question d

In [7]:
movie_review_0 = np.nan_to_num(movie_review, nan = 0) #impute missing values with 0

#Creating a "quick-SVD" function
def svd_reconstruct(M, k):    
    u_full, s_full , vh_full = np.linalg.svd(M, full_matrices = False)
    
    #at the end we only return a subset of u,s,v_transpose - this subset is dictated by k: # of non-zero singular values
    u = u_full[:,0:k] # (m x k) matrix - slicing only k-columns
    s = s_full[0:k] # (k x k) square matrix - selecting k non-zero singular values
    vh = vh_full[0:k,:] # (k x n) matrix - slicing only k-rows
    
    return u,s,vh

#Getting the SVD with k=19
u19,s19,vh19 = svd_reconstruct(movie_review_0,k=19)

#### Rank-19 reconstruction (with zero padding)
Based on my rank-19 reconstruction, the estimate of user 45 for the movie Jumanji is <ins>about 1.56.</ins>

In [8]:
movie_review_reconstruct = u19 @ np.diag(s19) @ vh19
print("Estimate of user 45 for Jumanji = " + str(movie_review_reconstruct[43][1])) #review of user 45 of movie Jumanji 

Estimate of user 45 for Jumanji = 1.5636012343776666


Furthermore, `difference_matrix` represents the difference for true (non-missing) values between this rank-19 reconstruction and the original `movie_review` matrix (i.e. $x_{true}-x_{recon}$). We can then compute the average difference, which is about <ins>1.71.</ins>


In [9]:
difference_matrix = movie_review - movie_review_reconstruct

N = np.sum(~np.isnan(difference_matrix)) 
#total number of values 2670*2153                   = 5748510
#total number of only non-missing (w/o nans) values = 1120381

print("Total number of true non-missing values (N) = " + str(N))
print("Avg difference (for true non-missing values) between review & reconstructed matrix = " 
      + str(math.sqrt(np.nansum(difference_matrix**2)/N)))

Total number of true non-missing values (N) = 1120381
Avg difference (for true non-missing values) between review & reconstructed matrix = 1.7076311829456958


## Question e
#### Imputing missing (nan) values via column-mean padding

In [10]:
col_mean = np.nanmean(movie_review,axis=0) #mean of each column
row_col_coord = np.where(np.isnan(movie_review)) #first element tells us the row where there is a nan-values
                                                    #second element tells us the column where there is a nan-values
                                                    #together, each pair forms (row,column) coord of each nan-values
row_col_coord

(array([   0,    0,    0, ..., 2669, 2669, 2669], dtype=int64),
 array([   3,    4,    5, ..., 2150, 2151, 2152], dtype=int64))

In [11]:
movie_review_colpad = movie_review.copy()
movie_review_colpad[row_col_coord] = np.take(col_mean,indices = row_col_coord[1])

#movie_review[row_column_coord]: extract out only nan-values in matrix movie_review to fill in: 4628129 nan values like N above
#np.take extracts out value from a particular array (col_mean in this case) based on certain indices
## we can fill in going row by row (row-major order) - each row stopping at column of nan (using 2nd element of row_col_coord)

#### Rank-19 reconstruction (with column-mean padding)
Based on my rank-19 reconstruction, the estimate of user 45 for the movie Jumanji is <ins>about 2.41</ins>.

In [12]:
#Getting the SVD with k=19
u19_col,s19_col,vh19_col = svd_reconstruct(movie_review_colpad,k=19)
movie_review_reconstruct_2 = u19_col @ np.diag(s19_col) @ vh19_col
print("Estimate of user 45 for Jumanji = " + str(movie_review_reconstruct_2[43][1])) #review of user 45 of movie Jumanji 

Estimate of user 45 for Jumanji = 2.41212644534551


For this rank-19 reconstruction (with column-mean padding), the average difference in true (non-missing) values (i.e. $x_{true}-x_{recon}$) is <ins>about 0.76</ins>.

In [13]:
difference_matrix_2 = movie_review - movie_review_reconstruct_2

print("Total number of true non-missing values (N) = " + str(N))
print("Avg difference (for true non-missing values) between review & reconstructed matrix = " 
      + str(math.sqrt(np.nansum(difference_matrix_2**2)/N)))

Total number of true non-missing values (N) = 1120381
Avg difference (for true non-missing values) between review & reconstructed matrix = 0.7580686426265324


# Staff-Graded Section
## Question f
#### Imputing missing (nan) values via row-mean padding

In [14]:
row_mean = np.nanmean(movie_review,axis=1)
movie_review_rowpad = movie_review.copy()

#we'll use the row_col_coord already created above
movie_review_rowpad[row_col_coord] = np.take(row_mean,indices = row_col_coord[0])

#### Rank-19 reconstruction (with row-mean padding)
Based on my rank-19 reconstruction, the estimate of user 45 for the movie Jumanji is <ins>about 2.32<ins>.

In [15]:
#Getting the SVD with k=19
u19_row,s19_row,vh19_row = svd_reconstruct(movie_review_rowpad,k=19)
movie_review_reconstruct_3 = u19_row @ np.diag(s19_row) @ vh19_row
print("Estimate of user 45 for Jumanji = " + str(movie_review_reconstruct_3[43][1])) #review of user 45 of movie Jumanji 

Estimate of user 45 for Jumanji = 2.3178785730722566


For this rank-19 reconstruction (with column-mean padding), the average difference in true (non-missing) values (i.e. $x_{true}-x_{recon}$) is <ins>about 0.77<ins>.

In [16]:
difference_matrix_3 = movie_review - movie_review_reconstruct_3

print("Total number of true non-missing values (N) = " + str(N))
print("Avg difference (for true non-missing values) between review & reconstructed matrix = " 
      + str(math.sqrt(np.nansum(difference_matrix_3**2)/N)))

Total number of true non-missing values (N) = 1120381
Avg difference (for true non-missing values) between review & reconstructed matrix = 0.7733516498877452


## Question g

#### Meaning of each imputation method
**Zero-padding**: In this method, any missing values are replaced by 0. By doing so, it's equivalent to making the assumption that if a user doesn't rate a movie, then it is probably because they wouldn't enjoy watching it anyway. Consequently, 0 - the lowest possible rating score - is given. Furthermore, assigning 0 means we're strictly focusing on the true (observed) value. However, this impacts our matrix reconstruction more (see more below in the comparison section).

**Column-mean padding**: In this method, any missing value <ins>for a given film</ins> us replaced by the average score (across all 2670 users) <ins>of said film</ins>. We're assuming that most people who watch film A (but didn't) would probably end up with opinions ranging somewhere in-between the already existing scores. Only a minor few would have extreme opinions, and these users probably have already rated film A had they felt so strongly about the move. Consequently, it makes sense to fill the "silent majority" with a metric that represents "in-between opinions": the arithmetic mean.

**Row-mean padding**: Similar to column-mean padding, this method replace missing values with an arithmetic mean, but across 2153 film ratings <ins>of the same user</ins>. Here, we instead assume that for any given person, their missing scores should just be "in-between" because if they had felt more strongly about a particular film, they would have left a rating already.

#### Comparison
Based on the average difference metric ($x_{true}-x_{recon}$), zero-padding is least effective at reducing the difference, followed by row-mean, then column-mean (although for column-mean and row-mean the metrics differ only by about 0.015). Specifically: 1.707 > 0.773 > 0.758.

In context, this makes sense because assigning 0 to all missing values is quite harsh - especially when the lowest singular rating is only 0.5 (identified in question a). No matter how terrible a movie is, people would be reluctant to give it a straight 0. And given that an indidual user have 2000+ film to rate, it is certain that we would have more 0 values than real values across the matrix if we used zero-padding (in fact the ratio of missing values to real values is 4:1). Consequently, our reconstructed matrix is "dragged" by the zero, impacting the average difference.

#### Alternative method - the sample average
Instead of calculating the average for each row or column, we can just use the sample average across the entire dataset. We calculate the sample average to be 3.297 (see below), and then assign 3.297 to every missing values. By doing so, we are again trying to represent the "silent majority" (instead of using an extreme value of 0) but this time regardless of movies and users. 

Because we're aggregating over the entire dataset, we may lose some nuance individual differences (in mean rating) between movies or users. Some movies may be too good and deserve a mean score significantly higher than 2.5, and vice versa. 

For all the above reasons, I expect this alternative method to be better than zero-padding but not as good as column-mean or row-mean padding. Indeed, the average difference (see below) is slightly higher (0.807 > 0.773 > 0.758).

In [17]:
avg = np.nansum(movie_review)/N

movie_review_alt = np.nan_to_num(movie_review, nan = avg) #impute missing values with 0
u_alt,s_alt,vh_alt = svd_reconstruct(movie_review_alt,k=19)
movie_review_recon_alt = u_alt @ np.diag(s_alt) @ vh_alt
difference_matrix_alt = movie_review - movie_review_recon_alt

print("Sample average = " + str(avg))
print("Avg difference (for true non-missing values) for alternative method = " 
      + str(math.sqrt(np.nansum(difference_matrix_alt**2)/N)))

Sample average = 3.297080189685473
Avg difference (for true non-missing values) for alternative method = 0.8069912138996118


## Question h
#### Non-negative Matrix Factorization (NMF) with K=19 components

In [18]:
NMF_model = NMF(n_components=19, init = 'random', random_state = 0, max_iter = 200) #create NMF model object

U_19 = NMF_model.fit_transform(movie_review_colpad) #Ut.shape = (2670, 19): this is transformed movie_review w/ 19 components
Vt_19 = NMF_model.components_ #Vt.shape = (19, 2153) for 19 NMF components (V_transpose)

C:\Users\tranq\anaconda3\lib\site-packages\sklearn\decomposition\_nmf.py:1090: ConvergenceWarning: Maximum number of iterations 200 reached. Increase it to improve convergence.
  warnings.warn("Maximum number of iterations %d reached. Increase it to"


#### `movie_review` reconstruction (with column-mean padding) using 19 NMF components
In this 19-component NMF-based reconstruction, the estimate of user 45 for the movie Jumanji is <ins>about 2.39</ins>.

In [19]:
movie_review_reconstruct_19 = U_19 @ Vt_19
print("Estimate of user 45 for Jumanji = " + str(movie_review_reconstruct_19[43][1])) #review of user 45 of movie Jumanji 

Estimate of user 45 for Jumanji = 2.3880948645121913


In this 19-component NMF-based reconstruction, the average difference in true (non-missing) values (i.e. $x_{true}-x_{recon}$) is <ins>about 0.77</ins>.

**Comparison to SVD-based average difference**: The average difference is <ins>almost identical</ins> to that found using SVD. The column-mean SVD has an average difference of 0.76. The row-mean SVD has an average difference 0.77. If we're splitting hairs, then this NMF has a worse average difference than column-mean SVD (0.768 > 0.758) and a better average difference than row-mean SVD (0.768 < 0.773).

In [20]:
difference_matrix_19 = movie_review - movie_review_reconstruct_19

print("Total number of true non-missing values (N) = " + str(N))
print("Avg difference (for true non-missing values) between review & reconstructed matrix = " 
      + str(math.sqrt(np.nansum(difference_matrix_19**2)/N)))

Total number of true non-missing values (N) = 1120381
Avg difference (for true non-missing values) between review & reconstructed matrix = 0.7683549412326072


## Question i

#### Reason for SVD:
The left and right matrices from SVD are orthogonal, and the columns of U and V are <ins>linearly independent (i.e. globally orthonormal bases)</ins>. Because of this, we can achieve a low-rank approximation and avoid multicollinearity issue (i.e. movies correlated with each other). In a review dataset like this one where there are many product columns (movies) with somewhat overlapping characteristics (genres), we can get a very high-dimensional matrix. Consequently, the above-mentioned benefit of SVD is quite useful. 

#### Reason for NMF:
An objective advantage NMF has over SVD is that it is more interpretable - and that is very true for this dataset. The NMF pick
different parts of the dataset, and as a result our data would lie very close to a specific axis (i.e. dimension). We can visualize each data point as if it belongs to 1 single (or some combination of) dimension(s). If we are able to define our dimensions as separate categories, we'd be able to interpret the NMF components meaningfully. This is perfect for our `movie_review` because it already has a well-defined set of categories: genres.

$\implies$For this reason, <ins>I'd recommend using NMF</ins>, especially if we want to test different combinations of genres as separate dimensions (i.e. Comedy/Drama vs. Horror/Thriller) rather than just 19 singular genre (my intial approach).

## Question j
#### Non-negative Matrix Factorization (NMF) with K=5 components

In [21]:
NMF_model_5 = NMF(n_components=5, init = 'random', random_state = 0, max_iter = 200) #create NMF model object

U_5 = NMF_model_5.fit_transform(movie_review_colpad) 
Vt_5 = NMF_model_5.components_ 

#Reconstruct
movie_review_reconstruct_5 = U_5 @ Vt_5

C:\Users\tranq\anaconda3\lib\site-packages\sklearn\decomposition\_nmf.py:1090: ConvergenceWarning: Maximum number of iterations 200 reached. Increase it to improve convergence.
  warnings.warn("Maximum number of iterations %d reached. Increase it to"


#### Recommender system - Top 1% 
For the recommendations, let's first find the movies in the 99th perecentile using the reconstructed ratings of user1462. There are 22 movies in the "top 1%". 

Additionally, I filtered out the movies that user1462 had already seen/rated - which are "The Usual Suspects", "I Love Trouble", "South Park", and "An Andalusian Dog". So we're down to 18 movies.

In [22]:
user1462_rec = movie_review_reconstruct_5[1460] #user1462
top_1 = np.where(user1462_rec > np.quantile(user1462_rec,0.99))[0] #indices where a movie is in top 1% for user1462
top_movie, top_genre, top_score = [],[],[]

#For-loop to check and store data for movies that meet the criteria
for i in top_1:
    movie = movie_review_names[i]
    genre = movie_review_genre[i]
    score = user1462[i]
    
    if math.isnan(score):
        top_movie.append(movie), top_genre.append(genre), top_score.append(user1462_rec[i])

#Create data frame containing movies recommended
rec_data = pd.DataFrame({'Movie Recommended':top_movie,
                         'Genre':top_genre,
                         'Rating Reconstruct':top_score})
rec_data = rec_data.sort_values(by=['Rating Reconstruct'],ascending=False).reset_index(drop=True)
rec_data

Movie Recommended                                 Genre  \
0                  Before Sunset (2004)                         Drama|Romance   
1                          K-PAX (2001)          Drama|Fantasy|Mystery|Sci-Fi   
2   Nine to Five (a.k.a. 9 to 5) (1980)                          Comedy|Crime   
3                 Reindeer Games (2000)                       Action|Thriller   
4                    My Dog Skip (1999)                        Children|Drama   
5     The Falcon and the Snowman (1985)                  Crime|Drama|Thriller   
6                        Gothika (2003)                       Horror|Thriller   
7                  Wild at Heart (1990)  Crime|Drama|Mystery|Romance|Thriller   
8    Slums of Beverly Hills, The (1998)                          Comedy|Drama   
9            Tales from the Hood (1995)                   Action|Crime|Horror   
10                      Red Heat (1988)                                Action   
11          Under the Tuscan Sun (2003)                  Comedy|Drama|Romance   
12                      Sideways (2004)                  Comedy|Drama|Romance   
13                Good Girl, The (2002)                          Comedy|Drama   
14          Gorillas in the Mist (1988)                                 Drama   
15                 Dreamers, The (2003)                                 Drama   
16          To Have and Have Not (1944)  Adventure|Drama|Romance|Thriller|War   
17               Day of the Dead (1985)                Horror|Sci-Fi|Thriller   

    Rating Reconstruct  
0             4.067384  
1             3.976258  
2             3.936378  
3             3.910926  
4             3.903575  
5             3.902192  
6             3.896596  
7             3.869991  
8             3.866889  
9             3.865597  
10            3.862120  
11            3.854656  
12            3.848307  
13            3.847080  
14            3.846345  
15            3.845918  
16            3.845425  
17            3.841003

**$\implies$** The above dataset orders the movies by their respective reconstructed ratings. <ins>I'd recommend "Before Sunset", "K-PAX", and "Nine to Five" as the top three to user1462.</ins> 

**Evaluation**: The 2nd and 3rd highest movie ("K-PAX" and "Nine to Five") do not align with my expectations because their respective genres (Mystery/Scifi and Comedy) are not even in the set of genre (see below) belonging to the 5-rated movies that user1462 originally had (question b). Because genre of a movie is important to a person's likings, recommend "K-PAX" and "Nine to Five" may backfire. Therefore, let's see if we can find a movie that is in the "top 1%" and also caters to user1462 preferences.

#### Additional Recommendations
First, I created a set that contains all 11 genres belonging to user1462's favorite movies.

In [23]:
genre_rec = set()

for i in index_review_1462:
    genre = movie_review_genre[i]
    genre_list = genre.split('|')
    for genre in genre_list:
        genre_rec.add(genre)

print(genre_rec)

{'Adventure', 'Thriller', 'War', 'Children', 'Fantasy', 'Action', 'Sci-Fi', 'Drama', 'Crime', 'Musical', 'Romance'}


Let' filter for any "top 1%" movie that shares these genre. The condition is strict: if a movie has more than one genre, all of those genres must be in common with this set.

In [24]:
top_movie, top_genre, top_score = [],[],[]

#For-loop to check and store data for movies that meet the criteria
for i in top_1:
    movie = movie_review_names[i]
    genre = movie_review_genre[i]
    score = user1462[i]
    genre_list = genre.split('|')
    
    if all(genre in genre_rec for genre in genre_list) and math.isnan(score): 
        top_movie.append(movie), top_genre.append(genre), top_score.append(user1462_rec[i])
        #filter for only movies with all genre in common w/ 5-rated list and user1462 has not yet seen

#Create data frame containing movies recommended
rec_data = pd.DataFrame({'Movie Recommended':top_movie,
                         'Genre':top_genre,
                         'Rating Reconstruct':top_score})
rec_data = rec_data.sort_values(by=['Rating Reconstruct'],ascending=False).reset_index(drop=True)
rec_data

Movie Recommended                                 Genre  \
0               Before Sunset (2004)                         Drama|Romance   
1              Reindeer Games (2000)                       Action|Thriller   
2                 My Dog Skip (1999)                        Children|Drama   
3  The Falcon and the Snowman (1985)                  Crime|Drama|Thriller   
4                    Red Heat (1988)                                Action   
5        Gorillas in the Mist (1988)                                 Drama   
6               Dreamers, The (2003)                                 Drama   
7        To Have and Have Not (1944)  Adventure|Drama|Romance|Thriller|War   

   Rating Reconstruct  
0            4.067384  
1            3.910926  
2            3.903575  
3            3.902192  
4            3.862120  
5            3.846345  
6            3.845918  
7            3.845425

$\implies$ Now, "K-PAX" and "Nine to Five" has been replaced by "Reindeer Games" and "My Dog Skip" - both of which are immediately below in the original ranking. 

#### Conclusion: 
I'd still recommend **<ins>"Before Sunset", "K-PAX", and "Nine to Five" as the top three</ins>** first, but I'd also propose <ins>"Before Sunset", "Reindeer Games" and "My Dog Skip" as an alternative</ins> top three because their categories are all in-common with user1462's original top-rated films.

## Question k
#### Soft imputation
First, let's implement the soft imputation algorithm using a 100-iteration for-loop:

In [25]:
#Iteration t=0: initialize intermediate matrix Y with NaN = 0
Y_next = np.nan_to_num(movie_review, nan = 0)
lambda_pen = 100

for iteration in range(0,100):
    u5,s5,vh5 = svd_reconstruct(Y_next,k=5)
    
    #Applying soft imputation using penalty lambda = 100
    for i in range(len(s5)):
        
        if s5[i] > lambda_pen:
            s5[i] = s5[i] - lambda_pen
        else:
            s5[i] = 0
    #Reconstruct Z^(t+1) for next iteration
    Z_next = u5 @ np.diag(s5) @ vh5
    
    #Iteration (t+1): fill missing values in movie_review w/ recon values in Z_next above
    next_iter = pd.DataFrame(movie_review).fillna(pd.DataFrame(Z_next))
    Y_next = np.array(next_iter,dtype=float)

Now, in the final iteration (100), our last reconstructed matrix - `Z_next` - will be the one we used to find top three movie to recommend to user45.

#### Recommender system - Top 3% 
Again, like in question j, let's first find the movies in the 97th perecentile using the reconstructed ratings of user45. For this user, I expanded my search to the "top 3%" because the 99th percentile ("top 1%") only yielded 2 movies.

There are 83 movies in the "top 3%". However, after filtering out the movies that user45 had already seen/rated, we're left with 18 movies.

In [26]:
Z_final = Z_next #the latest Z_next represents our final reconstructed matrix that will be used to recommend movies
user45_rec = Z_final[43] #user1462
top_3 = np.where(user45_rec > np.quantile(user45_rec,0.97))[0] #indices where a movie is in top 1% for user1462
top_movie, top_genre, top_score = [],[],[]

#For-loop to check and store data for movies that meet the criteria
for i in top_3:
    movie = movie_review_names[i]
    genre = movie_review_genre[i]
    score = user45[i] #this is original score not reconstructed score
    
    if math.isnan(score):
        top_movie.append(movie), top_genre.append(genre), top_score.append(user45_rec[i])

#Create data frame containing movies recommended
rec_data = pd.DataFrame({'Movie Recommended':top_movie,
                         'Genre':top_genre,
                         'Rating Reconstruct':top_score})
rec_data = rec_data.sort_values(by=['Rating Reconstruct'],ascending=False).reset_index(drop=True)
rec_data

Movie Recommended                           Genre  \
0          Ghost and Mrs. Muir, The (1947)           Drama|Fantasy|Romance   
1             2001: A Space Odyssey (1968)          Adventure|Drama|Sci-Fi   
2                                 M (1931)        Crime|Film-Noir|Thriller   
3                             Laura (1944)         Crime|Film-Noir|Mystery   
4                           Rebecca (1940)  Drama|Mystery|Romance|Thriller   
5            Dead Alive (Braindead) (1992)           Comedy|Fantasy|Horror   
6                    Apocalypse Now (1979)                Action|Drama|War   
7                  Harold and Maude (1971)            Comedy|Drama|Romance   
8               Killing Fields, The (1984)                       Drama|War   
9   American Werewolf in London, An (1981)          Comedy|Horror|Thriller   
10               Better Off Dead... (1985)                  Comedy|Romance   
11                      Being There (1979)                    Comedy|Drama   
12                       Local Hero (1983)                          Comedy   
13        Femme Nikita, La (Nikita) (1990)   Action|Crime|Romance|Thriller   
14                   Police Academy (1984)                    Comedy|Crime   
15                     Saving Grace (2000)                          Comedy   
16                 To Catch a Thief (1955)  Crime|Mystery|Romance|Thriller   
17                Somewhere in Time (1980)                   Drama|Romance   

    Rating Reconstruct  
0             3.585019  
1             3.550058  
2             3.471073  
3             3.470487  
4             3.452290  
5             3.438308  
6             3.429554  
7             3.423441  
8             3.418505  
9             3.380248  
10            3.372851  
11            3.342717  
12            3.338226  
13            3.326349  
14            3.318873  
15            3.318026  
16            3.312605  
17            3.306100

**$\implies$** The above dataset orders the movies by their respective reconstructed ratings. <ins>I'd recommend "The Ghost and Mrs. Muir", "2001: A Space Odyssey", and "M" as the top three to user45.</ins> 

**Evaluation**: Only the 3rd highest movie ("M") do not align with my expectations because two of its genre (Film-Noir and Thriller) are not even in the set of genre (see below) belonging to the 5-rated movies that user45 originally had (question b). Because genre of a movie is important to a person's likings, let's see if we can find another movie that is in the "top 3%" and also caters to user45 preferences.

#### Additional Recommendations
Like above, I created a set that contains all 7 genres belonging to user45's favorite movies.

In [27]:
genre_rec = set()

for i in index_review_5:
    genre = movie_review_genre[i]
    genre_list = genre.split('|')
    for genre in genre_list:
        genre_rec.add(genre)

print(genre_rec)

{'Adventure', 'Drama', 'Fantasy', 'Comedy', 'Crime', 'Sci-Fi', 'Romance'}


Let' filter for any "top 3%" movie that shares these genre. The condition is strict: if a movie has more than one genre, all of those genres must be in common with this set.

In [28]:
top_movie, top_genre, top_score = [],[],[]

#For-loop to check and store data for movies that meet the criteria
for i in top_3:
    movie = movie_review_names[i]
    genre = movie_review_genre[i]
    score = user45[i]
    genre_list = genre.split('|')
    
    if all(genre in genre_rec for genre in genre_list) and math.isnan(score): 
        top_movie.append(movie), top_genre.append(genre), top_score.append(user45_rec[i])
        #filter for only movies with all genre in common w/ 5-rated list and user1462 has not yet seen

#Create data frame containing movies recommended
rec_data = pd.DataFrame({'Movie Recommended':top_movie,
                         'Genre':top_genre,
                         'Rating Reconstruct':top_score})
rec_data = rec_data.sort_values(by=['Rating Reconstruct'],ascending=False).reset_index(drop=True)
rec_data

Movie Recommended                   Genre  Rating Reconstruct
0  Ghost and Mrs. Muir, The (1947)   Drama|Fantasy|Romance            3.585019
1     2001: A Space Odyssey (1968)  Adventure|Drama|Sci-Fi            3.550058
2          Harold and Maude (1971)    Comedy|Drama|Romance            3.423441
3        Better Off Dead... (1985)          Comedy|Romance            3.372851
4               Being There (1979)            Comedy|Drama            3.342717
5                Local Hero (1983)                  Comedy            3.338226
6            Police Academy (1984)            Comedy|Crime            3.318873
7              Saving Grace (2000)                  Comedy            3.318026
8         Somewhere in Time (1980)           Drama|Romance            3.306100

$\implies$ We end up with "Harold and Maude" replacing "M". 

#### Conclusion: 
I'd still recommend **<ins>"The Ghost and Mrs. Muir", "2001: A Space Odyssey", and "M" as the top three</ins>** first, and I'd also propose <ins>"Harold and Maude" as an alternative to "M"</ins> because its categories are all in-common with user45's original top-rated films.